File: solutions/haskell/food-chain/1/src/FoodChain.hs
module FoodChain (song) where

import Data.List (intercalate)
import Data.Char (toLower)
import Text.Printf (printf)

data Animal = Fly
            | Spider
            | Bird
            | Cat
            | Dog
            | Goat
            | Cow
            | Horse
            deriving (Show, Enum)

showLower :: Show a => a -> String
showLower = map toLower . show

song :: String
song = intercalate "\n" (map verse [Fly .. Horse])

verse :: Animal -> String
verse animal = unlines (swallowed animal : rhyme animal : maybeChain animal)
  where maybeChain Fly = []
        maybeChain Horse = []
        maybeChain animal = chain animal ++ [rhyme Fly]

swallowed :: Animal -> String
swallowed animal = printf "I know an old lady who swallowed a %s." (showLower animal)

chain :: Animal -> [String]
chain animal = zipWith whySwallow predators prey
  where predators = reverse [Fly .. animal]
        prey = tail predators

whySwallow :: Animal -> Animal -> String
whySwallow predator prey = printf "She swallowed the %s to catch the %s%s." (showLower predator) (showLower prey) (describe prey)
  where describe Spider = " that wriggled and jiggled and tickled inside her"
        describe _ = ""

rhyme :: Animal -> String
rhyme Fly = "I don't know why she swallowed the fly. Perhaps she'll die."
rhyme Spider = "It wriggled and jiggled and tickled inside her."
rhyme Bird = "How absurd to swallow a bird!"
rhyme Cat = "Imagine that, to swallow a cat!"
rhyme Dog = "What a hog, to swallow a dog!"
rhyme Goat = "Just opened her throat and swallowed a goat!"
rhyme Cow = "I don't know how she swallowed a cow!"
rhyme Horse = "She's dead, of course!"

File: solutions/go/food-chain/1/food_chain.go
package foodchain

import "strings"

const testVersion = 2

func Verse(v int) string {
	verse := "I know an old lady who swallowed " + first[v]
	if 8 <= v {
		return verse
	}
	for i := v; 0 <= i; i-- {
		verse += refrain[i]
	}

	return verse
}

func Verses(begin, end int) string {
	var verses = []string{}
	for v := begin; v <= end; v++ {
		verses = append(verses, Verse(v))
	}
	return strings.Join(verses, "\n\n") 
}

func Song() string {
	return Verses(1, 8)
}

var first = []string{
	"", 
	"a fly.\n",
	"a spider.\nIt wriggled and jiggled and tickled inside her.\n",
	"a bird.\nHow absurd to swallow a bird!\n",
	"a cat.\nImagine that, to swallow a cat!\n",
	"a dog.\nWhat a hog, to swallow a dog!\n",
	"a goat.\nJust opened her throat and swallowed a goat!\n",
	"a cow.\nI don't know how she swallowed a cow!\n",
	"a horse.\nShe's dead, of course!",
}

var refrain = []string{
	"", 
	"I don't know why she swallowed the fly. Perhaps she'll die.",
	"She swallowed the spider to catch the fly.\n",
	"She swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\n",
	"She swallowed the cat to catch the bird.\n",
	"She swallowed the dog to catch the cat.\n",
	"She swallowed the goat to catch the dog.\n",
	"She swallowed the cow to catch the goat.\n",
}